GitUI/GitUIExtensions.cs

```diff
@@ -184,25 +182,21 @@ public static IEnumerable<TreeNode> AllNodes(this TreeNodeCollection nodes)
             }
         }
 
-        public static void InvokeAsync(this Control control, Action action)
+        public static Task InvokeAsync(this Control control, Action action)
         {
-            InvokeAsync(control, _ => action(), null);
+            return InvokeAsync(control, _ => action(), null);
         }
 
-        public static void InvokeAsync(this Control control, SendOrPostCallback action, object state)
+        public static async Task InvokeAsync(this Control control, SendOrPostCallback action, object state)
         {
-            SendOrPostCallback checkDisposedAndInvoke = (s) =>
-            {
-                if (!control.IsDisposed)
-                {
-                    action(s);
-                }
-            };
+            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
 
-            if (!control.IsDisposed)
+            if (control.IsDisposed)
             {
-                UISynchronizationContext.Post(checkDisposedAndInvoke, state);
+                return;
             }
+
+            action(state);
         }
 
```

Comment:
I would not make this method async. Perhaps its name is of poor choice.

```csharp
public static void InvokeOnUIThread(this Control control, SendOrPostCallback action, object state)
{
    ThreadHelper.JoinableTaskFactory.RunAsync(async () => 
    {
        if (control.IsDisposed)
        {
            return;
        }
        await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
        if (control.IsDisposed)
        {
            return;
        }
        action(state);
    });
}
```

Reply:
This is what taskschedulers are for.  TaskScheduler.FromCurrentSynchronizationContext().  Store the result of that call in a global  because if you are in a different thread than ui it won't work.

Reply:
https://msdn.microsoft.com/en-us/library/system.threading.tasks.taskscheduler(v=vs.110).aspx?cs-save-lang=1&cs-lang=csharp#code-snippet-2

Reply:
@jbialobr This method is clearly performing an asynchronous operation. I have not removed all fire-and-forget cases, but a future cleanup operation will remove the remaining ones. It could be renamed to `InvokeOnUIThreadAsync` if you prefer.

@vbjay I have attempted to remove the use of non-default task schedulers as part of this pull request. Switching to the UI thread is now accomplished by of `SwitchToMainThreadAsync` inside an asynchronous method. Attempting to use task schedulers falls somewhere between unnecessary complexity and risk of deadlock. If you see any stragglers let me know.

Reply:
This change changes behavior from throwing exceptions on the UI thread to swallowing exceptions.
Compare https://github.com/gitextensions/gitextensions/commit/e1dcf888c5c6fcfa27dfa54986f1725d10e12a77 with https://github.com/gitextensions/gitextensions/commit/50e520951816fbbae41e7c145796019aaec8a683

Reply:
Why recreate the wheel? Tasks already have continuewith. We have control.invoke. There is no need to create a system that complicates and makes things not the normal way of things. If I see tasks I know how to make it run on the GUI. I know how to continue on the GUI thread.

Reply:
> This change changes behavior from throwing exceptions on the UI thread to swallowing exceptions.

This is somewhat intentional. Exceptions will no longer get thrown on a synchronization context (see next paragraph). However, we can handle exceptions for fire-and-forget style methods in any way you wish. Do you have a description of how you want these handled? Is this type of exception common in existing code, or is this something that can be implemented in a follow-up cleanup pull request?

There are *many* places in the code violating the rule that asynchronous operations cannot be started in an open-ended manner. These will need to be cleaned up in a later pass, and we are working on [analyzer enforcement](https://github.com/dotnet/roslyn-analyzers/issues/1590) to help prevent the problem from occurring again.

> Why recreate the wheel?

The .NET Tasks implementation is very good at many things, but it has some serious limitations when it comes to two specific areas:

1. Applications that have a dedicated UI thread
2. Applications that mix synchronous and asynchronous code

In the presence of the above, applications are prone to failures with one of the following root causes:

1. Deadlock involving the UI thread
2. Logic errors due to reentrancy during a blocking wait

The vs-threading library, approach, and analyzers make it substantially more difficult for errors in these categories to manifest at runtime. If this isn't already the new normal for applications that have a GUI thread, it will be soon. The logic behind the library is already hardened, and performance is proven at the scale of Visual Studio.

Reply:
>Do you have a description of how you want these handled?

Exceptions should be propagated to users (handled by the Report a buf form). Currently it is done by NBug attached to Application.ThreadException. These are exceptions mainly caused by a programmer mistake like NPE or IndexOutOfBounds.

Reply:
@jbialobr I believe the original behavior is now preserved.

Reply:
Thank you @sharwell. I verified that it works as previously.